src/antlr/QueryStructureVisitor.ts

```diff
@@ -0,0 +1,522 @@
+
+    // to be consumed later
+    this.relationsStack.push(this.currentRelation);
+    this.currentRelation = this.currentRelation.parent ?? new QueryRelation(this.getNextRelationId());
+    return result;
+  }
+
+  /**
+   * Processing set operations.
+   */
+  visitQueryTermDefault(ctx: QueryTermDefaultContext): Result {
+    // reinit column seq as we will repeat the same columns in subsequent queries
+    this.currentRelation.columnIdSeq = 0;
+    // clear relations for each queryTermDefault because it's individual query
+    this.currentRelation.relations = new Map();
+    return this.visitChildren(ctx);
+  }
+
+  visitRegularQuerySpecification(ctx: RegularQuerySpecificationContext): Result {
+    // process FROM first to capture all available relations
+    let result = ctx.fromClause()?.accept(this) ?? this.defaultResult();
+
+    // then process all remaining clauses
+    ctx.children?.forEach(c => {
+      if (!(c instanceof FromClauseContext)) {
+        result = this.aggregateResult(result, c.accept(this));
+      }
+    });
+
+    return result;
+  }
+
+  // processes table/CTE/correlated subquery references
+  visitTableName(ctx: TableNameContext): Result {
+    const multipartTableName = ctx
+      .multipartIdentifier()
+      .errorCapturingIdentifier()
+      .map(v => common.stripQuote(v.identifier()));
+
+    const strictId = ctx.tableAlias().strictIdentifier();
+    const alias = (strictId !== undefined
+      ? common.stripQuote(strictId)
+      : multipartTableName[multipartTableName.length - 1]
+    ).name;
+
+    if (multipartTableName.length == 1) {
+      const cte = this.currentRelation.findCTE(multipartTableName[0]);
+      if (cte !== undefined) {
+        // found relation as CTE
+        this.currentRelation.relations.set(alias, cte);
+        return this.defaultResult();
+      }
+
+      const rel = this.currentRelation.findRelation(multipartTableName[0]);
+      if (rel !== undefined) {
+        // found relation as correlated sq
+        if (multipartTableName[0].name != alias) {
+          this.currentRelation.relations.set(alias, rel);
+        }
+        return this.defaultResult();
+      }
+    }
+
+    const tablePrimary = common.tablePrimaryFromMultipart(multipartTableName.map(v => v.name));
+
+    const metadata = this.getTable(tablePrimary);
+    const columns = metadata?.columns.map(c => new Column(c.id, c.id, undefined, c.data)) || [];
+
+    const relation = new TableRelation(
+      this.getNextRelationId(),
+      tablePrimary,
+      columns,
+      this.currentRelation,
+      this.rangeFromContext(ctx),
+      metadata?.table.data
+    );
+
+    this.currentRelation.relations.set(alias, relation);
+    this.onRelation?.(relation, alias);
+
+    return this.defaultResult();
+  }
+
+  // processes CTE
+  visitNamedQuery(ctx: NamedQueryContext): Result {
+    const result = this.visitChildren(ctx);
+
+    // expecting query relation to be in stack
+    const relation = this.relationsStack.pop();
```

Comment:
Right now, both `relationStack` and `currentRelation` are modeling the "stack of relations". It seems the `relationsStack` here is only needed to access "the most recent popped relation". To use only `currentRelation` for this, an additional utility method can be used:
```typescript
const relation = this.scoped(() => this.visitChildren(ctx));
```
where `scoped` creates a new `QueryRelation` as a child of the current one, performs the child visiting, resets the original relation (i.e. "pops" the child relation) and returns the created child relation that now has all the infos from the subquery in it.

Reply:
I though about this approach. In this case we would need to handle all places where query term is used to introduce scope. And there are many such places which we could want to skip for now. Though probably `relationStack` can be replaced by singular variable as we don't use relations other than the top.

Reply:
Ah, you're right. We would need to cover all cases, which isn't ideal. I like the single variable suggestion though to only handle the most recent subquery.